### ISSUE TITLE
[sanitizers] [ASAN] False negative on global OOB

### ISSUE THREAD

**Issue by pskrgag** [llvm/llvm-project#108287]
Consider following code:

```c
char global[100];

void foo(void)
{
   global[101] = 10;
}
```

I'd expect to see ASAN report in such case, since there is obvious global OOB access. Suddenly, llvm does not insert any instrumentation for such access on x86, but inserts it on arm64. I tracked down it to `AddressSanitizer::isSafeAccess` function which optimizes away store instrumentation, since it think that access is in bounds. 

This is caused by the fact that `global` var on x86 ends up having align 16, but on arm64 it has align 1. After aligning up 100 to 16, access is indeed becomes in-bound, but it's not really obvious. 

Alignment difference is caused by `LargeArrayMinWidth` being equal to 128 on x86 (even with `-mno-sse`).

Is it expected behavior? Since from my perspective it seems a bit odd...

(reproduced on trunk https://godbolt.org/z/vKd3fjWYK)

**Comment by pskrgag** [comment]
CC: @vitalybuka 

I guess?

**Comment by vitalybuka** [comment]
Looks like a bug

**Comment by pskrgag** [comment]
If you have an idea where to look futher, I could try to fix it myself... The thing I don't really understand is how to get "real" alignment in `AddressSanitizer::isSafeAccess`. It's not possible to assume natural alignment, since we should respect `aligned` attrs.

I guess, at ir level this information is basially lost. Maybe `ObjectSizeOffsetVisitor::compute` should not take alignments into account and return real sizes (which seems to comformt language standarts, since OOB for over-aligned object is still UB)

**Comment by vitalybuka** [comment]
I am not sure if alignment has anything to do so.
In IR we have `@test = dso_local global { [100 x i8], [60 x i8] } zeroinitializer, comdat, align 32, !dbg !0`
and in assembly `.zero   160`

`60` are redzone and suppose to be poisoned.

The questing is what is `uint64_t Size` in :isSafeAccess`?
I would expect it should be 100.

**Comment by pskrgag** [comment]
> The questiong is what is uint64_t Size in :isSafeAccess`?
> I would expect it should be 100.

But it's actually 112, because initial alignment of global array is 16 (you can see alignment difference on x86 and arm64 here https://godbolt.org/z/aqosMTvMf).

So following leads to 112:

```
AddressSanitizer::isSafeAccess()
  ObjectSizeOffsetVisitor::compute()
    ObjectSizeOffsetVisitor::computeImpl()
     ObjectSizeOffsetVisitor::computeValue()
        ObjectSizeOffsetVisitor::visitGlobalVariable()
          return SizeOffsetAPInt(align(Size, GV.getAlign()), Zero);

```
On the last line `Size` is 100, but `GV.getAlign()` is 16. And `isSafeAccess` return true, since 101 < 112
